src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -485,21 +484,42 @@ private void AddSemanticRange(SyntaxNode node, int semanticKind)
             var childNodes = node.ChildNodes();
             if (childNodes.Count == 0)
             {
-                var content = node.GetContent();
-                var lines = content.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
                 var charPosition = range.Start.Character;
-                for (var i = 0; i < lines.Length; i++)
+                var lineStartAbsoluteIndex = node.SpanStart - charPosition;
+                for (var i = range.Start.Line; i <= range.End.Line; i++)
                 {
-                    var startPosition = new Position(range.Start.Line + i, charPosition);
-                    var endPosition = new Position(range.Start.Line + i, charPosition + lines[i].Length);
+                    var startPosition = new Position(i, charPosition);
+
+                    // NOTE: GetLineLength includes newlines but we don't report tokens for newlines so
+                    // need to account for them.
+                    var lineLength = source.Lines.GetLineLength(i);
+
+                    // For the last line, we end where the syntax tree tells us to. For all other lines, we end at the
+                    // last non-newline character
+                    var endChar = i == range.End.Line
+                       ? range.End.Character
+                       : GetLastNonWhitespaceCharacterOffset(source, lineStartAbsoluteIndex, lineLength);
+
+                    // Make sure we move our line start index pointer on, before potentially skipping out on the loop
+                    lineStartAbsoluteIndex += lineLength;
+                    charPosition = 0;
+
+                    // Blank line. The less than check is important because we could have seen two newline characters, but
+                    // be in a situation where we have an LF file with multiple blank lines, so endChar would be -1. Still
+                    // a blank line though :)
+                    if (endChar <= 0)
```

Comment:
❓ would it make more sense to have `GetLastNonWhitespaceCharacterOffset` not return a negative number?  Will comment below on proposed change

Reply:
Good call, thank you! Couldn't see the forrest for the trees :)